app/dependencies.py

```diff
@@ -15,29 +18,58 @@
 x_api_key_scheme = APIKeyHeader(name="x-api-key")
 
 
+@dataclass
 class AcaPyAuth:
     token: str
     role: Role
+    wallet_id: str = None
 
-    def __init__(self, *, role: "Role", token: str) -> None:
-        self.role = role
-        self.token = token
+
+@dataclass
+class AcaPyAuthVerified(AcaPyAuth):
+    wallet_id: str
 
 
 def acapy_auth(auth: str = Depends(x_api_key_scheme)):
     if not "." in auth:
         raise HTTPException(401, "Unauthorized")
 
-    [role_str, token] = auth.split(".", maxsplit=1)
+    try:
+        [role_str, token] = auth.split(".", maxsplit=1)
 
-    role = Role.from_str(role_str)
+        role = Role.from_str(role_str)
+    except:
+        raise HTTPException(401, "Unauthorized")
 
     if not role:
         raise HTTPException(401, "Unauthorized")
 
     return AcaPyAuth(role=role, token=token)
 
 
+def acapy_auth_verified(auth: AcaPyAuth = Depends(acapy_auth)):
+    if auth.role.is_admin:
+        if auth.token != auth.role.agent_type.x_api_key:
+            raise HTTPException(403, "Unauthorized")
+
+        wallet_id = "admin"
+    else:
+        try:
+            # Decode JWT
+            token_body = jwt.decode(
+                auth.token, ACAPY_MULTITENANT_JWT_SECRET, algorithms=["HS256"]
+            )
+        except jwt.InvalidTokenError:
+            raise HTTPException(403, "Unauthorized")
+
+        wallet_id = token_body.get("wallet_id")
```

Comment:
why do we use .get here?

Reply:
To not error out when the property doesn't exist. .get() returns None by default so we can handle the wallet not exists a few lines below.

Could change it to a try/except but feels quite verbose to me and still doesn't capture the property exists but value is None (as that won't throw an eror)